app/src/main/java/org/luxoft/sdl_core/ClassicBtHandler.java

```diff
@@ -0,0 +1,375 @@
+
+    // Constants that indicate the current connection state
+    public static final int STATE_NONE = 0;       // we're doing nothing
+    public static final int STATE_LISTEN = 1;     // now listening for incoming connections
+    public static final int STATE_CONNECTING = 2; // now initiating an outgoing connection
+    public static final int STATE_CONNECTED = 3;  // now connected to a remote device
+
+    public static final String TAG = ClassicBtHandler.class.getSimpleName();
+
+    public static synchronized ClassicBtHandler getInstance(Context context) {
+        if (instance == null) {
+            instance = new ClassicBtHandler(context);
+        }
+        return instance;
+    }
+
+    private ClassicBtHandler(Context context) {
+        this.context = context;
+        // Register for broadcasts when a device is discovered.
+        IntentFilter filter = new IntentFilter(BluetoothDevice.ACTION_FOUND);
+        context.registerReceiver(mReceiver, filter);
+        mState = STATE_NONE;
+    }
+
+    public void DoDiscovery() {
+        Log.i(TAG, "If we're already discovering, stop it");
+        //
+        if (mBtAdapter.isDiscovering()) {
+            mBtAdapter.cancelDiscovery();
+        }
+        Log.i(TAG, "Request discover from BluetoothAdapter");
+        // Request discover from BluetoothAdapter
+        mBtAdapter.startDiscovery();
+    }
+
+    public void disconnect() {
+        Log.d(TAG, "Closing bluetooth handler...");
+        context.unregisterReceiver(mReceiver);
+    }
+
+    /**
+     * The BroadcastReceiver that listens for discovered devices and changes the title when
+     * discovery is finished
+     */
+    private synchronized void connect(BluetoothDevice device) {
+        Log.d(TAG, "connect to: " + device);
+
+        // Cancel any thread attempting to make a connection
+        if (mState == STATE_CONNECTING) {
+            if (mConnectThread != null) {
+                mConnectThread.cancel();
+                mConnectThread = null;
+            }
+        }
+
+        // Cancel any thread currently running a connection
+        if (mConnectedThread != null) {
+            mConnectedThread.cancel();
+            mConnectedThread = null;
+        }
+
+        // Start the thread to connect with the given device
+        mConnectThread = new ConnectThread(device);
+        mConnectThread.start();
+    }
+
+    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            String action = intent.getAction();
+
+            // When discovery finds a device
+            if (BluetoothDevice.ACTION_FOUND.equals(action)) {
+                // Get the BluetoothDevice object from the Intent
+                BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
+                String deviceName = device.getName();
```

Comment:
@ychernysheva add null check here for `device` as it can be not included into the intent